### ISSUE TITLE
PV/Psuedo-Irradience Model/Labeller

### ISSUE THREAD

**Issue by jacobbieker** [openclimatefix/psuedo-pv-labeller#2]
This is to summarize a meeting about this project and how to define it better:

Questions:
1. What is PV Labelling?
     - The models would label/forecast a psuedo-irradience/PV value, normalized by pvlib's clear sky irradience. To get a specific generation for a tilt/orientation there would need to be a small model that takes this pseudo-irradience and maps it to that configuration. The output from the labelling/forecasting model is not actual irradience or PV generation, but something that downstream models can learn to use to improve their predictions.
3. What Model do we want to use?
     - Simple model first, probably 3D CNN to have temporal features
4. What are the inputs? All are history (if forecasting), and both history+future (if labelling)
     - Patch of data in Lat/Lon
     - NWP, if available
     - Satellite imagery, if available
     - PV Generation, if available
     - Topographic map
     - Sun position
5. What are the outputs?
     - Latent variable (between 0 and 1)
     - Grid, same as input Lat/Lon, so that it scales well
7. How will the outputs get used?
     - Forecasting: Another input feature to site-level models
     - Labelling: Another data source to train all models on, as well as possibly doing hindcasts for potential solar sites

Diagram of data flow:
![image](https://user-images.githubusercontent.com/7170359/221835589-7b627ef0-6d61-4368-946c-053751853272.png)

Next steps:
  - @jacobbieker will start on 2a, then work backwards to 1 if 2a helps downstream models.
  - Test how much current models accuracy increases as number of sites trained on increases. (i.e. 10,100,1000) and see if the accuracy starts plateauing (to see if more sites, real or fake, will actually improve model performance)

**Comment by jacobbieker** [comment]
@simlmx @peterdudfield @dantravers this summary look good?

**Comment by dantravers** [comment]
Looks good to me, thanks @jacobbieker!  My comments: 
I think it's correct to say that the model you describe could be the same, or a similar model used for 1 and 2a, if I understand correctly?  Might be worth making that clear. 
And for the outputs in (4), the latent variable could be a vector of variables, rather than just one.

**Comment by jacobbieker** [comment]
Thanks! Updated it

**Comment by jacobbieker** [comment]
For pixels that have multiple sites in them, how should we handle it? The first approach I was going to go with is just take a random one and ignore the others, but would it be worth trying to 'blend' the different sites in a pixel together?

**Comment by dantravers** [comment]
I would firstly differentiate the sites based on orientation - I.e. use orientation as an additional dimension (together with lat, lon).  So we fit either an irradiance based on lat, lon, orient.  Or we could generate a vector of irradiance figures - so we can predict different orientations with the same pixel.

But the problem still exists, as if you say bucket into groups of orientation - E, SE, S, SW, W, ... - you will still get duplicates in each bucket.  I would suggest clustering the sites in each bucket and taking the centroid?

**Comment by jacobbieker** [comment]
Yeah, that makes sense for how to normalize it. But for the 2 points that are in the same pixel, there could easily be sub-pixel clouds and such that means they won't have the same value, right?

**Comment by simlmx** [comment]
Yes that's a good point. I would try to use a resolution that makes those not too frequent and treat what's left as noise.

In any case we should look at what the data looks like to understand in what ways values at the same location differ.

**Comment by jacobbieker** [comment]
The satellite pixels are between 3-9km wide, while NWP is 1-2km, so we can't go more precise than that, unfortunately. 

But yeah, that sounds good.

**Comment by simlmx** [comment]
We can and should be more precise than the satellite or NWP data.

Our *ground truth*, the PV data, has an arbitrary resolution (because we have very precise coordinates for the sites).  Some of the *features* (satellite, NWP) that we use don't have a high resolution but I don't think that's an issue. On the other hand, the features that we can compute from the neighbouring PV sites has a high resolution, and should definitely be leveraged. I think this is where (high resolution PV data) most of the gains from this approach will come from, at least for short term predictions in regions dense with PVs, mainly because of the higher resolution.

**Comment by simlmx** [comment]
@dantravers This is a good question and that's why I think the first step is to verify those assumptions!

I think you are right with your example. Maybe there is a way to use the different irradiance values (direct, indirect, etc.) to do a better normalization (that would also work for you example), but I'm not sure.

An alternative is to ignore those extreme cases them during training. By training on non-ambiguous targets, we would get a more stable model, and I think that model would still be useful for downstream tasks (e.g. as features in PV site forecast). In the downstream task, the model will learn to not use those feature in the extreme cases, if they are not useful.

**Comment by dantravers** [comment]
Similar to the "ignoring extreme cases..." could be to just use sites with a 180 orientation (+/-10), and compare the accuracy to using the full set?

**Comment by simlmx** [comment]
Yeah why not, anything that makes the task we are trying to learn unambiguous. Then we can see where it is best to cut (time of day, orientation of panels, etc.).

We could also try training with averages for ambiguous cases, who knows maybe learning the average is actually what is best for the downstream task.

One thing to note is that even if we end up removing some data points when we create our **targets**, I don't think that we want to remove anything when we use the PV data as **features**.
